### DISCUSSION TITLE
Markdoc language server extension fails to run in VS Code

### DISCUSSION THREAD

**Discussion by crcorbett** [markdoc/markdoc#460] [Q&A]
### What happened?

Hi!

I am trying to make use of the **Markdoc language server** extension within VS Code to provide validation, autocomplete etc. Unfortunately, I can't seem to be able to get the extension to perform correctly despite following the [documentation](https://github.com/markdoc/language-server) provided by Markdoc.

I have created a simple repo (available here: https://github.com/crcorbett/tilt-legal-site) to test out Markdoc's functionality. I have installed the Markdoc language server extension within my VS Code workspace. I have created a simple Markdoc schema, a basic `post.mdoc`, and confirmed that Markdoc is working correctly locally.

When I select a file ending in `.mdoc` I can see that the Markdoc extension is being activated as it appears within the status bar of VS Code (see screenshot **below**).
![Screenshot 2023-10-23 at 1 24 30 pm](https://github.com/markdoc/markdoc/assets/45161689/177e63ef-29e9-4379-98d4-57af94d0a88a)

However, I don't seem to be able to get any features of the extension to work. When typing within the `.mdoc` file no autocomplete hints or validations appear.

Furthermore, when activating the `Control the markdoc language server` action from the command palette the following error message appears (see screenshots **below**):
![Screenshot 2023-10-23 at 1 28 52 pm](https://github.com/markdoc/markdoc/assets/45161689/14bd7553-f2d4-4319-bb01-813eecbda7a0)
![Screenshot 2023-10-23 at 12 34 22 pm](https://github.com/markdoc/markdoc/assets/45161689/35adc68b-b68e-4613-a31a-2307728f4269)

Similarly, the `Display a rendered preview of Markdoc content` action does not succeed — though no error message appears here i.e. it fails silently.

----

Appreciating that this is an open-source project, I have tried to exhaust the possibilities for where I could've introduced a bug before raising this issue. I've taken a look through the previously closed issues that relate the Markdoc language server (e.g. #428) and attempted to follow the advice given in these.

Additionally, I have also tried cloning other repos that appear to have correct configurations to test the extension including:
- https://github.com/storj/docs/
- https://github.com/icerpc/icerpc-docs

Unfortunately, I am still having the same issue with these other repos.

This leads me to believe that there may be some issue with the configuration of my VS Code application distinct from any issues with Markdoc or the Markdoc language server. I have tested disabling all other extensions, which was unsuccessful and also had a colleague test out these repos on their VS Code install to no avail.

I have provided extracts of our Markdoc schema and config below for your reference.

Please let me know if there is any further detail I can provide to assist with debugging this issue.

Your assistance is greatly appreciated! 😃

-----

## Markdoc schema
https://github.com/crcorbett/tilt-legal-site/blob/main/src/app/blog/config.markdoc.js
```js
import { Heading, Callout, Paragraph } from "./(components)";


export const config = {
  nodes: {
    paragraph: {
      render: "Paragraph",
    },
    heading: {
      render: "Heading",
      attributes: {
        level: { type: Number },
      },
    },
  },
  tags: {
    callout: {
      render: "Callout",
      attributes: {
        title: {
          type: String,
          default: "", // default heading for callout component
        },
      },
    },
  },
};


export const components = {
  Paragraph,
  Heading,
  Callout,
};
```

## Markdoc config (Extension)
https://github.com/crcorbett/tilt-legal-site/blob/ef9139e034cfb4b6b235b443e992456ab5280707/markdoc.config.json

```js
[
  {
    "id": "Blog",
    "path": "src/app/blog/(articles)/",
    "schema": {
      "path": "src/app/blog/config.markdoc.js",
      "type": "esm",
      "property": "config",
      "watch": true
    }
  }
]
```

### To reproduce

Clone our `tilt-legal-site` test repository:
```bash
gh repo clone crcorbett/tilt-legal-site
```
Then install necessary packages and run locally:
```bash
pnpm i
pnpm run dev
```
### VS Code
Version: 1.83.1 (Universal)
Commit: f1b07bd25dfad64b0167beb15359ae573aecd2cc
Date: 2023-10-10T23:46:55.789Z
Electron: 25.8.4
ElectronBuildId: 24154031
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Darwin arm64 22.5.0

### Markdoc language support extension
v0.0.11

### Version

0.3.4

**Comment by rpaul-stripe** [comment]
When the language server loads a schema, it literally attempts to import it, so it has to conform with the requirements imposed by the Node.js runtime for imports.

On the first line of your schema, you have something like this: `import {...} from "./(components)"`, which is a directory import that attempts to load an `index.ts` file, which is not valid in a pure ES module. The schema file itself also has a `.js` extension, which also prevents it from being loaded as an ES module. After changing the extension to `.mjs` and removing both the line that imports the React components as well as the component export at the bottom I was able to get it to load successfully in the language server.

In projects that use TypeScript and couple React components with the schema, what I'd recommend is adding a build step of some kind to generate a pure schema output. You can have a TypeScript file somewhere that just imports the schema part of the configuration, and then do a standalone build of that entry point. I typically do this with esbuild in my projects.

I'm going to convert this issue to a discussion, but I will explore some options for exposing schema loading errors and see if there's something that I can do to make these kinds of issues less opaque to the end user. There should be an easy way to see when there is a schema loading exception.

**Reply by ocupe** [reply]
@rpaul-stripe would you mind sharing how you build the config file with esbulid? I have no luck so far.
Also a link to a repo that does this would be helpful :)

**Reply by johno-stripe** [reply]
Hey @Ocupe , I recently ran across this in my own project and can share my build strategy!  My project didn't already have an esbuild configured, so I just did this with Typescript.  Here's the dedicated `tsconfig.markdoc.json` I set up for this purpose:

```json
{
    "include": [
        "./src/path/to/my/markdocSchema.ts"
    ],
    "compilerOptions": {
        "target": "esnext",
        "module": "CommonJS",
        "rootDir": "./src",
        "outDir": "./compiledSchema",
        "esModuleInterop": true,
        "forceConsistentCasingInFileNames": true,
        "strict": true,
        "noImplicitAny": true,
        "skipLibCheck": true
    }
}
```

My Markdoc schema file is importing from a bunch of others, and the way this is configured, most of that folder structure gets reproduced in the build.  This is my `markdoc.config.json` file:

```json
[
    {
        "id": "my-custom-markdoc",
        "path": "markdocFiles",
        "schema": {
            "path": "./compiledSchema/path/to/my/markdocSchema.js",
            "type": "node",
            "watch": true
        }
    }
]
```

Running `npx tsc --project ./tsconfig.markdoc.json` now produces a valid schema which gets picked up by the extension and validates in my editor.

I wouldn't say that this is the "best" approach, but it's been a good fit for my project!  Couple of caveats that my config assumes:
- CommonJS modules, so you'll need to tweak it if your `package.json` includes `"type": "module"`.
- Your Markdoc schema is the default export out of `markdocSchema.ts`

**Comment by Sporiff** [comment]
I'm experiencing the exact same issue using the [example setup from Astro](https://docs.astro.build/en/guides/integrations-guide/markdoc/#vs-code-editor-integration)
